tests/InferTypesFuzz.elm

```diff
@@ -0,0 +1,329 @@
+        |> Random.map (literal Float)
+
+
+boolExpr : Generator Canonical.LocatedExpr
+boolExpr =
+    Random.bool
+        |> Random.map (literal Bool)
+
+
+charExpr : Generator Canonical.LocatedExpr
+charExpr =
+    Random.int 0 0x0010FFFF
+        |> Random.map Char.fromCode
+        |> Random.map (literal Char)
+
+
+stringExpr : Generator Canonical.LocatedExpr
+stringExpr =
+    Random.int 0 0x0010FFFF
+        |> Random.list 10
+        |> Random.map (List.map Char.fromCode)
+        |> Random.map String.fromList
+        |> Random.map (literal String)
+
+
+unitExpr : Generator Canonical.LocatedExpr
+unitExpr =
+    Canonical.Unit
+        |> located
+        |> Random.constant
+
+
+literal : (a -> Literal) -> a -> Canonical.LocatedExpr
+literal wrap value =
+    value
+        |> wrap
+        |> Canonical.Literal
+        |> located
+
+
+listExpr : Type -> Generator Canonical.LocatedExpr
+listExpr elementType =
+    elementType
+        |> exprGenerator
+        |> Random.list 10
+        |> Random.map Canonical.List
+        |> Random.map located
+
+
+intToIntFunctionExpr : Generator Canonical.LocatedExpr
+intToIntFunctionExpr =
+    let
+        combine argument intPart =
+            Canonical.lambda argument <|
+                located <|
+                    Canonical.Plus
+                        (located <| Canonical.Argument argument)
+                        intPart
+
+        intSubExpr =
+            exprGenerator Type.Int
+    in
+    Random.map2 combine
+        -- TODO: Later we will need something better to avoid shadowing.
+        randomVarName
+        intSubExpr
+        |> Random.map located
+
+
+randomVarName : Generator VarName
+randomVarName =
+    let
+        starters =
+            "abcdefghijklnmoqprstuvwxyz"
+
+        others =
+            "ABCDEFGHIJKLMNOQPRSTUVQXYZ_0123456789"
+
+        all =
+            starters ++ others
+
+        charFrom string =
+            string
+                |> String.toList
+                |> List.map Random.constant
+                |> Random.choices (Random.constant 'a')
+
+        firstChar =
+            charFrom starters
+
+        rest =
+            Random.list 5 <| charFrom all
+    in
+    Random.map2 (::) firstChar rest
+        |> Random.map String.fromList
+        |> Random.map VarName
+
+
+dumpType : Type -> String
+dumpType type_ =
+    type_
+        |> Type.toString Type.emptyState
+        |> Tuple.first
+
+
+{-| An expression shrinker that preserves the inferred type.
+-}
+exprShrinker : Shrinker Canonical.LocatedExpr
+exprShrinker expr =
+    case Located.unwrap expr of
+        Canonical.Literal lit ->
+            lit
+                |> shrinkLiteral
+                |> Shrink.map Canonical.Literal
+                |> Shrink.map located
+
+        Canonical.Plus left right ->
+            expr
+                |> shrinkPlus left right
+
+        Canonical.List elements ->
+            elements
+                |> shrinkNonEmptyList exprShrinker
```

Comment:
Just checking my understanding... we're not shrinking `[1]` to `[]` because that would make the type not `List Int` but `List a`, which in itself isn't a problem (`List a` can be used where you need `List Int`) but our current testing solution doesn't currently have good functions to check that? (ie. we only have equality, which will say `List Int /= List a`)

ie. shrinking to empty list would have to be accompanied with a different operator in the test cases themselves (typechecks / is a subtype / unifies / ... instead of equality)

:question:

Reply:
That is exactly the reason.

Reply:
Let's write a comment here: something along the lines of "We don't want to shrink non-empty list to an empty one - that would change the type, eg. from `List Int` to `List a`"